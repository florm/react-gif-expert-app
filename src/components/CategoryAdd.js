import React, { useState } from 'react'
import PropTypes from 'prop-types'

export const CategoryAdd = ({setCategories}) => {

    //esto me tira un error:
    //Warning: A component is changing an uncontrolled input to be controlled. This is likely caused by the value changing from undefined to a defined value, 
    //which should not happen. Decide between using a controlled or uncontrolled input element for the lifetime of the component.
    //porque esta pasando de un estado undefined a un estado con valor
    //const [inputValue, setInputValue] = useState();

    //para evitar esto lo ponemos como un string vacio, para que no
    //sea undefined
    const [inputValue, setInputValue] = useState('');

    //con un estado inicializado esto no pasa
    //const [inputValue, setInputValue] = useState('Hola');
    

    const handleInputChange = (e) => {
        setInputValue(e.target.value);
    }

    const handleSubmit = (e) => {
        e.preventDefault(); //siempre evitar comportamiento por defecto
        if(inputValue.trim().length > 2){
            setCategories(cat => [inputValue, ...cat]);
        }
        setInputValue('');
    }

    return (
        <form onSubmit = { handleSubmit }>
           <input
                type="text"
                value = {inputValue}
                onChange = { handleInputChange }
                
           />
        </form>
    )

    
}

CategoryAdd.propTypes = {
    setCategories: PropTypes.func.isRequired
}





