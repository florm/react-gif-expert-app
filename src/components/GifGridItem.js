import React from 'react'
import PropTypes from 'prop-types'

export const GifGridItem = ({id, title, url}) => { //si lo leo asi parece que cuando llame a <GifGridItem/> lo hice usando 3 atributos distintos, tipo:
    //<GifGridItem id={id} title={title} url={url}/>
    
    return (
        <div className='card animate__animated animate__fadeIn'>
             <img src={ url } alt={ title }/>
             <p>{ title }</p>
        </div>
    )
}

GifGridItem.propTypes ={
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
}

