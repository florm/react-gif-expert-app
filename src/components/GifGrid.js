import React, { useEffect, useState } from 'react'
import { getGifs } from '../helpers/getGifs';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({category}) => {
    
    // const [images, setImages] = useState([]);
    const { data: images, loading } = useFetchGifs(category);
    
    // useEffect(()=>{

    //     getGifs(category).then(imgs => {
    //         setImages(imgs);
    //     });
    //     //o simplificando
    //     getGifs(category).then(setImages);

    // }, [category]) //los [] indican que la función se va a ejecutar 1 sola vez
    // //si agrego algo entre los [] indica que cuando lo que este ahi adentro cambie, entonces se vuelve a ejecutar. En este caso ponermos la category para que no de el warning,
    // //porque la categoria no va a cambiar nunca en este caso

    
    
    //con este metodo aca, cada vez que se renderice el componente se va a ejecutar
    //y por lo tanto se va a llamar a la pai por cada vez. 
    //lo comento, y lo uso en un useEffect
    //getGifs();
    
        
    return (
        <>
            <h3 className="animate__animated animate__fadeIn"> { category } </h3> 
            
            { loading && <p className = "animate__animated animate__flash" >Loading...</p>}
            <div className='card-grid'>
                {
                    images.map( img => {
                        //esta es una forma de mandar la prop img. Cada uno de los atributos se reciben en una única prop en GifGridItem
                        //si en GifGridItem yo quiero desestructurar puedo hacer {prop1, prop2} etc. Pero img corresponderia a prop1.
                            // return <GifGridItem key={img.id} img ={img}/>
                            //Entonces lo que necesito es 
                            //desestructurar a su vez img para poder recibir todos ssus componentes (id, title, url) por separado. Uso:
                            return <GifGridItem key={img.id} {...img}/>
                            //Es como si mandara cada una de las propiedad de img como una propiedad independiente. Lo de arriba quedaria como escribir esto:
                            // return <GifGridItem key={img.id} id={img.id} title={img.title} url={img.url}/>
                    })
                }
            
            </div>
        </>
    )
}
