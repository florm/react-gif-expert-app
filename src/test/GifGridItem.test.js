import {shallow} from 'enzyme'
import { GifGridItem } from '../components/GifGridItem'

describe('test para GifGridItem', () => {
    
    let wrapper
    
    test('se debe mostrar correctamente el GifGridItem', () => {
        wrapper = shallow( <GifGridItem/>)
        expect(wrapper).toMatchSnapshot()
    })
    
})
