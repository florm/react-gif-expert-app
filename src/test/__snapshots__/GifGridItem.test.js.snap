// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test para GifGridItem se debe mostrar correctamente el GifGridItem 1`] = `
<div
  className="card animate__animated animate__fadeIn"
>
  <img />
  <p />
</div>
`;
